run/idp-sql/cloud-sql.js

```diff
@@ -0,0 +1,178 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const Knex = require('knex');
+const { getCredConfig } = require('./secrets');
+const { logger } = require('./logging');
+
+const TABLE = process.env.TABLE || 'votes';
+let knex, credConfig;
+
+// Connection pooling config
+// See Cloud SQL sample https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/cloud-sql/postgres/knex
+// to learn more about this configuration
+const config = {
+  pool: {
+    max: 5,
+    min: 5,
+    acquireTimeoutMillis: 60000,
+  },
+  createTimeoutMillis: 30000,
+  idleTimeoutMillis: 600000,
+  createRetryIntervalMillis: 200,
+}
+
+// [START run_user_auth_sql_connect]
+/**
+* Connect to the Cloud SQL instance through UNIX Sockets
+*
+* @param {object} credConfig The Cloud SQL connection configuration from Secret Manager
+* @returns {object} Knex's PostgreSQL client
+*/
+const connectWithUnixSockets = async (credConfig) => {
+  const dbSocketPath = process.env.DB_SOCKET_PATH || "/cloudsql"
+  // Establish a connection to the database
+  return Knex({
+    client: 'pg',
+    connection: {
+      user: credConfig.DB_USER, // e.g. 'my-user'
+      password: credConfig.DB_PASSWORD, // e.g. 'my-user-password'
+      database: credConfig.DB_NAME, // e.g. 'my-database'
+      host: `${dbSocketPath}/${credConfig.CLOUD_SQL_CONNECTION_NAME}`,
+    },
+    ...config
+  });
+}
+// [END run_user_auth_sql_connect]
+
+// Method to connect locally on Windows
+const connectWithTcp = (credConfig) => {
+  // Extract host and port from socket address
+  const dbSocketAddr = process.env.DB_HOST.split(":") // e.g. '127.0.0.1:5432'
+  // Establish a connection to the database
+  return Knex({
+    client: 'pg',
+    connection: {
+      user: credConfig.DB_USER, // e.g. 'my-user'
+      password: credConfig.DB_PASSWORD, // e.g. 'my-user-password'
+      database: credConfig.DB_NAME, // e.g. 'my-database'
+      host: dbSocketAddr[0], // e.g. '127.0.0.1'
+      port: dbSocketAddr[1], // e.g. '5432'
+    },
+    ...config
+  });
+}
+
+
+/**
+* Connect to the Cloud SQL instance through TCP or UNIX Sockets
+* dependent on DB_HOST env var
+*
+* @returns {object} Knex's PostgreSQL client
+*/
+const connect = async () => {
+  if (!credConfig) credConfig = await getCredConfig();
+  if (process.env.DB_HOST) {
+    return connectWithTcp(credConfig);
+  } else {
+    return connectWithUnixSockets(credConfig);
+  }
+}
+
+/**
+* Insert a vote record into the database.
+*
+* @param {object} vote The vote record to insert.
+* @returns {Promise}
+*/
+const insertVote = async (vote) => {
+  if (!knex) knex = await connect();
+  await createTable(); // Create postgreSQL table if not found
```

Comment:
nit: I would suggest moving this to somewhere where it triggers once when the app starts (probably when you initialize your connection pool), instead of checking every time you need a statement.